Sources/Core/SecureStore/Cryptography.swift

```diff
@@ -0,0 +1,218 @@
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import Foundation
+
+typealias SecureStoreQuery = [String: Any]
+typealias KeyPair = (privateKey: SecKey, publicKey: SecKey)
+
+enum Cryptography {
+  static let service: String = "com.facebook.sdk.cryptography.\(Bundle.main.bundleIdentifier ?? "")"
+
+  @discardableResult
+  static func generateRSAKeyPair() throws -> KeyPair {
+    let query: SecureStoreQuery = [
+      SecureStoreKeys.attrKeyType: SecureStoreKeys.attrKeyTypeRSA,
+      SecureStoreKeys.attrKeySizeInBits: 2048,
+      SecureStoreKeys.privateKeyAttrs: [
+        SecureStoreKeys.attrIsPermanent: true,
+        SecureStoreKeys.attrApplicationTag: service as Any
+      ],
+      SecureStoreKeys.publicKeyAttrs: [
+        SecureStoreKeys.attrIsPermanent: true,
+        SecureStoreKeys.attrApplicationTag: service as Any
+      ]
+    ]
+
+    var potentialPublicKey: SecKey?
+    var potentialPrivateKey: SecKey?
+
+    let status = SecKeyGeneratePair(
+      query as CFDictionary,
+      &potentialPublicKey,
+      &potentialPrivateKey
+    )
+
+    guard status == noErr,
+      let privateKey = potentialPrivateKey,
+      let publicKey = potentialPublicKey
+      else {
+        throw CryptographyError.keyGenerationFailure
+    }
+
+    return (privateKey, publicKey)
+  }
+
+  static func deleteRSAKeyPair() throws {
+    let query: SecureStoreQuery = [
+      SecureStoreKeys.class: SecureStoreKeys.cryptographyKey,
+      SecureStoreKeys.attrApplicationTag: service,
+      SecureStoreKeys.attrKeyType: SecureStoreKeys.attrKeyTypeRSA,
+      SecureStoreKeys.returnReference: true
+    ]
+
+    let status = SecItemDelete(query as CFDictionary)
+
+    guard status == noErr || status == errSecItemNotFound else {
+      throw CryptographyError.unhandledError(status)
+    }
+  }
+
+  static func rsaPublicKey() throws -> SecKey {
+    let query: SecureStoreQuery = [
+      SecureStoreKeys.class: SecureStoreKeys.cryptographyKey,
+      SecureStoreKeys.attrApplicationTag: service,
+      SecureStoreKeys.attrKeyType: SecureStoreKeys.attrKeyTypeRSA,
+      SecureStoreKeys.attrKeyClass: SecureStoreKeys.attrKeyClassPublic,
+      SecureStoreKeys.returnReference: true
+    ]
+
+    var itemReference: AnyObject?
+    let status = SecItemCopyMatching(query as CFDictionary, &itemReference)
+
+    guard status == noErr || status == errSecSuccess else {
+      throw CryptographyError.publicKeyRetrievalFailure
+    }
+
+    // TODO: decide if we should create a new pair for them
+    guard status != errSecItemNotFound else {
+      throw CryptographyError.missingRSAPublicKey
+    }
+
+    // This allows for safe force-unwrapping below
+    guard CFGetTypeID(itemReference) == SecKeyGetTypeID() else {
+      throw CryptographyError.publicKeyInvalidType
+    }
+
+    // This is really Hinky but there is no ideal way to unwrap an optional `AnyObject`
```

Comment:
Perhaps a less ambiguous comment

Reply:
Fair. Do you have any ideas on how to phrase this? It's basically forcing me to force-unwrap because of a language bug. I guess I could just put that..

Reply:
Updated the message to hopefully be more clear that it's a defense of force-downcasting